From b7cb15cda1de0b91acf6fdc18d3e05c2bf332a61 Mon Sep 17 00:00:00 2001
From: csz <christopher.storah@leicabiosystems.com>
Date: Mon, 9 Dec 2024 15:00:32 +1100
Upstream-Status: Pending
Subject: [PATCH] integrated test pattern patch

---
 drivers/media/i2c/evdmoom1.c | 79 ++++++++++++++++++++++++++++++++++++
 1 file changed, 79 insertions(+)

diff --git a/drivers/media/i2c/evdmoom1.c b/drivers/media/i2c/evdmoom1.c
index 0cb2df30e..4e8d5d3b5 100644
--- a/drivers/media/i2c/evdmoom1.c
+++ b/drivers/media/i2c/evdmoom1.c
@@ -21,6 +21,7 @@
 #include <linux/mutex.h>
 #include <linux/regmap.h>
 #include <linux/regulator/consumer.h>
+#include <linux/bits.h>
 
 #include <media/media-entity.h>
 #include <media/v4l2-ctrls.h>
@@ -266,6 +267,7 @@
 #define AP1302_WARNING(n)			AP1302_REG_16BIT(0x6004 + (n) * 2)
 #define AP1302_SENSOR_SELECT			AP1302_REG_16BIT(0x600c)
 #define AP1302_SENSOR_SELECT_TP_MODE(n)		((n) << 8)
+#define AP1302_SENSOR_SELECT_TP_MODE_MASK	GENMASK(11, 8)
 #define AP1302_SENSOR_SELECT_PATTERN_ON		BIT(7)
 #define AP1302_SENSOR_SELECT_MODE_3D_ON		BIT(6)
 #define AP1302_SENSOR_SELECT_CLOCK_SEC		BIT(5)
@@ -275,6 +277,7 @@
 #define AP1302_SENSOR_SELECT_SENSOR_TP		(0U << 0)
 #define AP1302_SENSOR_SELECT_SENSOR_PRIM	(1U << 0)
 #define AP1302_SENSOR_SELECT_SENSOR_SEC		(2U << 0)
+#define AP1302_SENSOR_SELECT_SENSOR_MASK	GENMASK(1, 0)
 #define AP1302_SYS_START			AP1302_REG_16BIT(0x601a)
 #define AP1302_SYS_START_PLL_LOCK		BIT(15)
 #define AP1302_SYS_START_LOAD_OTP		BIT(12)
@@ -1490,6 +1493,72 @@ static int ap1302_set_mipi_t3_clk(struct ap1302_device *ap1302)
  * V4L2 Controls
  */
 
+enum {
+	AP1302_TP_MODE_DISABLED = 0,
+	AP1302_TP_MODE_FLAT_COLOR,
+	AP1302_TP_MODE_PSEUDO_RANDOM,
+	AP1302_TP_MODE_COLOR_BARS,
+	AP1302_TP_MODE_GREY_BARS,
+	AP1302_TP_MODE_PRBS1,
+	AP1302_TP_MODE_PRBS2,
+	AP1302_TP_MODE_PRBS3,
+	AP1302_TP_MODE_PRBS4,
+	AP1302_TP_MODE_V_STRIPES,
+	AP1302_TP_MODE_V_RAMP,
+	AP1302_TP_MODE_WALKING_ONES_10B,
+	AP1302_TP_MODE_WALKING_ONES_8B,
+	AP1302_TP_MODE_BW,
+};
+
+static const char * const tp_qmenu[] = {
+	"Disabled",
+	"Flat Color",
+	"Pseudo Random (Noise)",
+	"100% Color Bars",
+	"Fade to Grey Bars",
+	"PRBS1",
+	"PRBS2",
+	"PRBS3",
+	"PRBS4",
+	"Vertical Stripes",
+	"Vertical Ramp",
+	"Walking 1's (10bit)",
+	"Walking 1's (8bit)",
+	"Black and White",
+};
+
+static int ap1302_set_test_pattern(struct ap1302_device *ap1302, s32 pat)
+{
+	u32 val;
+	int ret;
+
+	ret = ap1302_read(ap1302, AP1302_SENSOR_SELECT, &val);
+	if (ret)
+		return ret;
+
+	if (pat == AP1302_TP_MODE_DISABLED) {
+		val &= ~AP1302_SENSOR_SELECT_SENSOR_MASK;
+		/* The sensor field in the sensor select register defaults to primary
+		 * sensor (0x1). Therefore, when disabling test pattern mode, restore
+		 * the value by setting it to the primary sensor
+		 */
+		val |= AP1302_SENSOR_SELECT_SENSOR_PRIM;
+		val &= ~AP1302_SENSOR_SELECT_PATTERN_ON;
+		val &= ~AP1302_SENSOR_SELECT_TP_MODE_MASK;
+		val |= AP1302_SENSOR_SELECT_TP_MODE(pat);
+	} else if (pat >= AP1302_TP_MODE_FLAT_COLOR && pat <= AP1302_TP_MODE_BW) {
+		val &= ~AP1302_SENSOR_SELECT_SENSOR_MASK;
+		val |= AP1302_SENSOR_SELECT_SENSOR_TP;
+		val |= AP1302_SENSOR_SELECT_PATTERN_ON;
+		val &= ~AP1302_SENSOR_SELECT_TP_MODE_MASK;
+		val |= AP1302_SENSOR_SELECT_TP_MODE(pat);
+	} else {
+		return -EINVAL;
+	}
+
+	return ap1302_write(ap1302, AP1302_SENSOR_SELECT, val, NULL);
+}
+
 static u16 ap1302_wb_values[] = {
 	AP1302_AWB_CTRL_MODE_OFF,	/* V4L2_WHITE_BALANCE_MANUAL */
 	AP1302_AWB_CTRL_MODE_AUTO,	/* V4L2_WHITE_BALANCE_AUTO */
@@ -2041,6 +2110,9 @@ static int ap1302_s_ctrl(struct v4l2_ctrl *ctrl)
 	case V4L2_CID_AP1302_STEREO_ORDER:
 		return ap1302_set_stereo_order(ap1302, ctrl->val);
 
+	case V4L2_CID_TEST_PATTERN:
+		return ap1302_set_test_pattern(ap1302, ctrl->val);
+
 	default:
 		dev_dbg(ap1302->dev, "Unknown control 0x%x\n",ctrl->id);
 		return -EINVAL;
@@ -2294,6 +2366,13 @@ static const struct v4l2_ctrl_config ap1302_ctrls[] = {
 		.max = 0xFFFF,
 		.step = 1,
 		.def = 0,
+	}, {
+		.ops = &ap1302_ctrl_ops,
+		.id = V4L2_CID_TEST_PATTERN,
+		.min = 0,
+		.max = (ARRAY_SIZE(tp_qmenu) - 1),
+		.def = 0,
+		.qmenu = tp_qmenu,
 	}
 };
 
-- 
2.45.2

