From b50f0f02bdcbf07b7060448df0ae39953c1c8950 Mon Sep 17 00:00:00 2001
From: csz <christopher.storah@leicabiosystems.com>
Date: Mon, 9 Dec 2024 17:13:59 +1100
Upstream-Status: Pending
Subject: [PATCH] fixing up test pattern control

---
 drivers/media/i2c/evdmoom1.c | 174 ++++++++++++-----------------------
 1 file changed, 59 insertions(+), 115 deletions(-)

diff --git a/drivers/media/i2c/evdmoom1.c b/drivers/media/i2c/evdmoom1.c
index 4e8d5d3b5..5ac706aee 100644
--- a/drivers/media/i2c/evdmoom1.c
+++ b/drivers/media/i2c/evdmoom1.c
@@ -515,6 +515,7 @@ struct ap1302_device {
 	bool use_vcid; // virtual channel
 	bool stereo_order;
 	bool primary_clk_rst_only;
+	u32 test_pattern;
 };
 
 static inline struct ap1302_device *to_ap1302(struct v4l2_subdev *sd)
@@ -552,19 +553,46 @@ static const struct ap1302_format_info supported_video_formats[] = {
 			 | AP1302_PREVIEW_OUT_FMT_FST_YUV_420,
 		.pixel_rate = AP1302_INITIAL_PIXEL_RATE,
 	},
-//#if defined(MEDIA_BUS_FMT_VYYUYY8_1X24) // MEDIA_BUS_FMT_VYYUYY8_1X24 for Xilinx only ?
 	{
 		.code = MEDIA_BUS_FMT_SRGGB12_1X12,
 		.out_fmt = AP1302_PREVIEW_OUT_FMT_FT_RAW12
 			 | AP1302_PREVIEW_OUT_FMT_FST_RAW_SENSOR,
 		.pixel_rate = AP1302_INITIAL_PIXEL_RATE,
 	},
-//#endif
 };
 
-static const struct ap1302_sensor_info ap1302_sensor_info_tpg = {
-	.model = "tpg",
-	.resolution = { 1920, 1080 },
+enum {
+	AP1302_TP_MODE_DISABLED = 0,
+	AP1302_TP_MODE_FLAT_COLOR,
+	AP1302_TP_MODE_PSEUDO_RANDOM,
+	AP1302_TP_MODE_COLOR_BARS,
+	AP1302_TP_MODE_GREY_BARS,
+	AP1302_TP_MODE_PRBS1,
+	AP1302_TP_MODE_PRBS2,
+	AP1302_TP_MODE_PRBS3,
+	AP1302_TP_MODE_PRBS4,
+	AP1302_TP_MODE_V_STRIPES,
+	AP1302_TP_MODE_V_RAMP,
+	AP1302_TP_MODE_WALKING_ONES_10B,
+	AP1302_TP_MODE_WALKING_ONES_8B,
+	AP1302_TP_MODE_BW,
+};
+
+static const char * const tp_qmenu[] = {
+	"Disabled",
+	"Flat Color",
+	"Pseudo Random (Noise)",
+	"100% Color Bars",
+	"Fade to Grey Bars",
+	"PRBS1",
+	"PRBS2",
+	"PRBS3",
+	"PRBS4",
+	"Vertical Stripes",
+	"Vertical Ramp",
+	"Walking 1's (10bit)",
+	"Walking 1's (8bit)",
+	"Black and White",
 };
 
 /* -----------------------------------------------------------------------------
@@ -1329,9 +1357,6 @@ static int ap1302_configure(struct ap1302_device *ap1302)
 	unsigned int data_lanes = ap1302->bus_cfg.bus.mipi_csi2.num_data_lanes;
 	int ret = 0;
 	u16 width_factor = ap1302_width_factor(ap1302);
-	//u8 sensor_link = ((!!ap1302->sensors[1].link_enabled)<<1) |
-	//		!!ap1302->sensors[0].link_enabled;
-	u8 sensor_link = 0;
 
 	u32 value = AP1302_PREVIEW_HINF_CTRL_SPOOF |
 			AP1302_PREVIEW_HINF_CTRL_MIPI_LANES(data_lanes);
@@ -1341,55 +1366,28 @@ static int ap1302_configure(struct ap1302_device *ap1302)
 
 	ap1302_write(ap1302, AP1302_PREVIEW_HINF_CTRL, value, &ret);
 
-	value =  AP1302_SENSOR_SELECT_TP_MODE(4) ;
+	value =  AP1302_SENSOR_SELECT_TP_MODE(ap1302->test_pattern) ;
 
-	switch(sensor_link)
-	{
-		case 0:
-			value |= AP1302_SENSOR_SELECT_PATTERN_ON |
-				AP1302_SENSOR_SELECT_SENSOR_TP;
-			break;
-		case 1:
-			value |= AP1302_SENSOR_SELECT_SENSOR_PRIM |
-			AP1302_SENSOR_SELECT_RESET_PRIM |
-			AP1302_SENSOR_SELECT_CLOCK_PRIM;
-			break;
-		case 2:
-			value |= AP1302_SENSOR_SELECT_SENSOR_SEC;
-			if (ap1302->primary_clk_rst_only) {
-				value |= AP1302_SENSOR_SELECT_RESET_PRIM |
-					AP1302_SENSOR_SELECT_CLOCK_PRIM;
-			}
-			else {
-				value |= AP1302_SENSOR_SELECT_RESET_SEC |
-					AP1302_SENSOR_SELECT_CLOCK_SEC;
-			}
-			break;
-		case 3:
-			if (ap1302->use_vcid)
-				value |= AP1302_SENSOR_SELECT_SENSOR_PRIM;
-			else
-				value |= ap1302->stereo_order ?
-					AP1302_SENSOR_SELECT_SENSOR_SEC :
-					AP1302_SENSOR_SELECT_SENSOR_PRIM ;
-
-			value |= AP1302_SENSOR_SELECT_RESET_PRIM |
-				AP1302_SENSOR_SELECT_CLOCK_PRIM |
-				AP1302_SENSOR_SELECT_MODE_3D_ON ;
-
-			if (!ap1302->primary_clk_rst_only)  {
-				value |= AP1302_SENSOR_SELECT_RESET_SEC |
-					AP1302_SENSOR_SELECT_CLOCK_SEC;
-			}
+	if (ap1302->test_pattern == AP1302_TP_MODE_DISABLED) {
+		dev_dbg(ap1302->dev, "Test pattern disabled - using primary sensor.\n");
+		value |= AP1302_SENSOR_SELECT_PATTERN_ON |
+			AP1302_SENSOR_SELECT_SENSOR_TP;
+	} else {
+		dev_dbg(ap1302->dev, "Test pattern %d\n", ap1302->test_pattern);
+		value |= AP1302_SENSOR_SELECT_TP_MODE(0) | AP1302_SENSOR_SELECT_SENSOR_PRIM |
+			AP1302_SENSOR_SELECT_RESET_PRIM | AP1302_SENSOR_SELECT_CLOCK_PRIM;
 	}
 
-	dev_dbg(ap1302->dev,"Sensor Select 0x%x\n",value);
+	dev_dbg(ap1302->dev,"Sensor Select = 0x%x\n",value);
 	ap1302_write(ap1302, AP1302_SENSOR_SELECT, value, &ret);
 
+	dev_dbg(ap1302->dev,"format.width = 0x%x\n", format->format.width);
 	ap1302_write(ap1302, AP1302_PREVIEW_WIDTH,
 		     format->format.width / width_factor, &ret);
+	dev_dbg(ap1302->dev,"format.height = 0x%x\n", format->format.height);
 	ap1302_write(ap1302, AP1302_PREVIEW_HEIGHT,
 		     format->format.height, &ret);
+	dev_dbg(ap1302->dev,"format.width = 0x%x\n", format->format.width);
 	ap1302_write(ap1302, AP1302_PREVIEW_OUT_FMT,
 		     format->info->out_fmt, &ret);
 	if (ret < 0)
@@ -1493,70 +1491,16 @@ static int ap1302_set_mipi_t3_clk(struct ap1302_device *ap1302)
  * V4L2 Controls
  */
 
-enum {
-	AP1302_TP_MODE_DISABLED = 0,
-	AP1302_TP_MODE_FLAT_COLOR,
-	AP1302_TP_MODE_PSEUDO_RANDOM,
-	AP1302_TP_MODE_COLOR_BARS,
-	AP1302_TP_MODE_GREY_BARS,
-	AP1302_TP_MODE_PRBS1,
-	AP1302_TP_MODE_PRBS2,
-	AP1302_TP_MODE_PRBS3,
-	AP1302_TP_MODE_PRBS4,
-	AP1302_TP_MODE_V_STRIPES,
-	AP1302_TP_MODE_V_RAMP,
-	AP1302_TP_MODE_WALKING_ONES_10B,
-	AP1302_TP_MODE_WALKING_ONES_8B,
-	AP1302_TP_MODE_BW,
-};
-
-static const char * const tp_qmenu[] = {
-	"Disabled",
-	"Flat Color",
-	"Pseudo Random (Noise)",
-	"100% Color Bars",
-	"Fade to Grey Bars",
-	"PRBS1",
-	"PRBS2",
-	"PRBS3",
-	"PRBS4",
-	"Vertical Stripes",
-	"Vertical Ramp",
-	"Walking 1's (10bit)",
-	"Walking 1's (8bit)",
-	"Black and White",
-};
+static int ap1302_get_test_pattern(struct ap1302_device *ap1302, s32 *pat)
+{
+	*pat = ap1302->test_pattern;
+	return 0;
+}
 
 static int ap1302_set_test_pattern(struct ap1302_device *ap1302, s32 pat)
 {
-	u32 val;
-	int ret;
-
-	ret = ap1302_read(ap1302, AP1302_SENSOR_SELECT, &val);
-	if (ret)
-		return ret;
-
-	if (pat == AP1302_TP_MODE_DISABLED) {
-		val &= ~AP1302_SENSOR_SELECT_SENSOR_MASK;
-		/* The sensor field in the sensor select register defaults to primary
-		 * sensor (0x1). Therefore, when disabling test pattern mode, restore
-		 * the value by setting it to the primary sensor
-		 */
-		val |= AP1302_SENSOR_SELECT_SENSOR_PRIM;
-		val &= ~AP1302_SENSOR_SELECT_PATTERN_ON;
-		val &= ~AP1302_SENSOR_SELECT_TP_MODE_MASK;
-		val |= AP1302_SENSOR_SELECT_TP_MODE(pat);
-	} else if (pat >= AP1302_TP_MODE_FLAT_COLOR && pat <= AP1302_TP_MODE_BW) {
-		val &= ~AP1302_SENSOR_SELECT_SENSOR_MASK;
-		val |= AP1302_SENSOR_SELECT_SENSOR_TP;
-		val |= AP1302_SENSOR_SELECT_PATTERN_ON;
-		val &= ~AP1302_SENSOR_SELECT_TP_MODE_MASK;
-		val |= AP1302_SENSOR_SELECT_TP_MODE(pat);
-	} else {
-		return -EINVAL;
-	}
-
-	return ap1302_write(ap1302, AP1302_SENSOR_SELECT, val, NULL);
+	ap1302->test_pattern = pat;
+	return 0;
 }
 
 static u16 ap1302_wb_values[] = {
@@ -2185,6 +2129,9 @@ static int ap1302_g_ctrl(struct v4l2_ctrl *ctrl)
 	case V4L2_CID_AP1302_STEREO_ORDER:
 		return ap1302_get_stereo_order(ap1302, &ctrl->val);
 
+	case V4L2_CID_TEST_PATTERN:
+		return ap1302_get_test_pattern(ap1302, &ctrl->val);
+
 	default:
 		return -EINVAL;
 	}
@@ -2336,8 +2283,7 @@ static const struct v4l2_ctrl_config ap1302_ctrls[] = {
 		.max = 1,
 		.step = 1,
 		.def = 0,
-	},
-	{
+	},{
 		.ops = &ap1302_ctrl_ops,
 		.id = V4L2_CID_HBLANK,
 		.name = "HBlank",
@@ -2346,8 +2292,7 @@ static const struct v4l2_ctrl_config ap1302_ctrls[] = {
 		.max = 0xFFFF,
 		.step = 1,
 		.def = 0,
-	},
-	{
+	}, {
 		.ops = &ap1302_ctrl_ops,
 		.id = V4L2_CID_VBLANK,
 		.name = "VBlank",
@@ -2356,8 +2301,7 @@ static const struct v4l2_ctrl_config ap1302_ctrls[] = {
 		.max = 0xFFFF,
 		.step = 1,
 		.def = 0,
-	},
-	{
+	}, {
 		.ops = &ap1302_ctrl_ops,
 		.id = V4L2_CID_PIXEL_RATE,
 		.name = "Pixel Rate",
-- 
2.45.2

